### PR TITLE
CA-201639: Add destination network validation before vm migrate

### PR DESCRIPTION
**Pull request by cheng-z**
Signed-off-by: chengz <cheng.zhang@citrix.com>

### CODE REVIEW

File: ocaml/xapi/xapi_vm_migrate.ml [hunk 1]
```diff
@@ -92,6 +92,14 @@ open Storage_interface
 open Listext
 open Fun
 
+let assert_network_support_migration ~__context ~vif_map ~remote =
+  let network_records = XenAPI.Network.get_all_records remote.rpc remote.session in
+  let network_refs = List.map (fun (_,record) ->  XenAPI.Network.get_by_uuid remote.rpc remote.session record.API.network_uuid) network_records in
```
Comment (simonjbeaumont) [inline ocaml/xapi/xapi_vm_migrate.ml]: isn't the first thing in this tuple the ref? You don't need this step.

File: ocaml/xapi/xapi_vm_migrate.ml [hunk 2]
```diff
@@ -92,6 +92,14 @@ open Storage_interface
...
+  List.iter (fun (vif, network) ->
+    let exist_network = List.filter (fun network_ref -> network_ref = network ) network_refs in
```
Comment (simonjbeaumont) [inline ocaml/xapi/xapi_vm_migrate.ml]: Please use List.exists instead of checking the length of List.filter.

**Comment by simonjbeaumont** [comment]
I think this logic should be done when the VIF->network map is resolved. Rather than as another additional pre-flight check. It's been placed alongside the check SR support but it's not the same class of check. That check was checking that the SR on the source had the necessary functionality (snapshotting) to make migration possible. Later on we resolve both the VDI->SR map and the VIF->Network map and I think the error handling should be there.

**Comment by thomassa** [comment]
Added the "needs updating" label based on Si's comment about when to do this validation.

File: ocaml/idl/datamodel.ml [hunk 3]
```diff
@@ -1016,6 +1016,8 @@ let _ =
     ~doc:"An HA statefile could not be created, perhaps because no SR with the appropriate capability was found." ();
   error Api_errors.vif_not_in_map [ "vif" ]
     ~doc:"This VIF was not mapped to a destination Network in VM.migrate_send operation" () ;
+  error Api_errors.network_not_found [ "network" ]
+    ~doc:"This network can not be found in the destination Networks in VM.migrate_send operation" () ;
```
Comment (thomassa) [inline ocaml/idl/datamodel.ml]: The words "can not" should be "cannot", but in any case I feel it is better to give errors names and descriptions that allow us to use them in other circumstances in the future: there's no need to limit this to VM.migrate_send.
```
error Api_errors.remote_network_not_found [ "network" ]
  ~doc:"The specified network does not exist on the remote pool." ();
```
or
```
error Api_errors.network_not_found [ "network" ]
  ~doc:"The specified network does not exist." ();
```

Comment (cheng-z) [inline ocaml/idl/datamodel.ml]: Thanks @thomassa 
I am now using option 2.
Thanks again.

**Comment by johnelse** [comment]
Agreed with @cheng-z that this should wait until after #2625

**Pull request closed without merging**